File: src/main/java/com/wowo/wowo/domain/wallet/repository/WalletRepository.java
package com.wowo.wowo.domain.wallet.repository;

import com.wowo.wowo.domain.wallet.entity.WalletAggregate;
import com.wowo.wowo.domain.wallet.valueobjects.WalletId;
import com.wowo.wowo.exception.NotFoundException;

import java.util.Optional;

/**
 * Domain repository interface for Wallet aggregate.
 * Defines the contract for wallet persistence operations.
 */
public interface WalletRepository {
    
    /**
     * Saves a wallet aggregate.
     * 
     * @param wallet the wallet to save
     * @return the saved wallet
     */
    WalletAggregate save(WalletAggregate wallet);
    
    /**
     * Finds a wallet by its ID.
     * 
     * @param walletId the wallet ID
     * @return the wallet if found
     */
    Optional<WalletAggregate> findById(WalletId walletId);
    
    /**
     * Finds a wallet by its ID or throws an exception.
     * 
     * @param walletId the wallet ID
     * @return the wallet
     * @throws NotFoundException if wallet not found
     */
    default WalletAggregate findByIdOrThrow(WalletId walletId) {
        return findById(walletId)
            .orElseThrow(() -> new NotFoundException("Wallet not found: " + walletId.getValue()));
    }
    
    /**
     * Checks if a wallet exists by its ID.
     * 
     * @param walletId the wallet ID
     * @return true if the wallet exists
     */
    boolean existsById(WalletId walletId);
    
    /**
     * Deletes a wallet by its ID.
     * 
     * @param walletId the wallet ID
     */
    void deleteById(WalletId walletId);
}

File: src/main/kotlin/com/wowo/wowo/contexts/wallet/domain/repository/WalletRepository.kt
package com.wowo.wowo.contexts.wallet.domain.repository

import com.wowo.wowo.contexts.wallet.domain.entity.Wallet
import com.wowo.wowo.contexts.wallet.domain.valueobject.WalletId
import com.wowo.wowo.shared.valueobject.Currency
import com.wowo.wowo.contexts.wallet.domain.valueobject.OwnerType

/**
 * Repository interface for Wallet aggregate (Domain layer)
 */
interface WalletRepository {
    fun save(wallet: Wallet): Wallet
    fun findById(id: WalletId): Wallet?
    fun findByOwnerIdAndOwnerType(ownerId: String, ownerType: OwnerType): List<Wallet>
    fun existsByOwnerIdAndOwnerType(ownerId: String, ownerType: OwnerType): Boolean
    fun existsByOwnerIdAndOwnerTypeAndCurrency(ownerId: String, ownerType: OwnerType, currency: Currency): Boolean
    fun delete(wallet: Wallet)
}

